Add recipe for syncthing

KeyWeeUsr opened the pull request:
### Brief summary of what the package does

This package attempts to port the browser client functionality into Emacs.

### Direct link to the package repository

https://github.com/KeyWeeUsr/emacs-syncthing

### Your association with the package

Maintainer

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X] My elisp byte-compiles cleanly
- [X] I've used `M-x checkdoc` to check the package's documentation strings
- [X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

progfolio commented:
Cool idea. A few notes:

You should use `defgroup` to define the syncthing group before assigning user options to it.
The syncthing group can be a subgroup of applications.


```emas-lisp
  (when (not syncthing-token)
```

This can be better expressed by using `(unless syncthing-token ...`.

If you `define-derived-mode` a syncthing-mode, you won't have to manually kill local variables via `syncthing--setup-buffer`. Enabling the major mode will do that. This also provides the user with a hook for when the mode is enabled and a dedicated keymap (vs the hard-coded `local-set-key` you have for `syncthing--newline` now).


```emacs-lisp
(defun syncthing--url (path)
  "Assemble full API url from PATH."
  (format "https://127.0.0.1:8384/%s" path))
```

This should be a user option.
Syncthing can be made to listen on a different port.


You have a number of functions concerned with propertizing text.
Instead, use `defface` to define faces.
This allows users to customize the look of the application and will significantly simplify your package.
Where possible, inherit from Emacs [standard faces](https://www.gnu.org/software/emacs/manual/html_node/emacs/Standard-Faces.html). When specifying a text size, it's better to use a relative size than hard-code a height value.


```
(defvar syncthing--fold-folders
  (list)
  "Tmp to hold IDs of folds.")
```

This can be initialized to `nil` to avoid the unnecessary function call.
The docstring can mention that it's a list. e.g. "List of fold IDs".

From a design perspective, you may want to declare these state variables via `defvar-local`.
That would allow running more than one client (useful if the user has more than one server running).

The auto-refresh functionality could be implemented as a minor mode.

KeyWeeUsr commented:
@progfolio Now that was a rather annoying, yet still worthwhile rewrite :sweat_smile: 
* `defgroup`s and even with subgroups for customization
* `(unless ...)`
* `(syncthing)` func to spawn a buffer and attach a major mode
* major derived `syncthing-mode` with local reference bound to buffer name, therefore allowing multiple sessions both with `(syncthing)` as a spawner and `(syncthing-mode)` as a "buffer hijacker"
* better state cleanup
* dedicated keymap
* customizable URL and other parts too
* `defface`d and `:inherit`ed from some present faces + relative `:height` (or at least doc says float = relative)
* Some of the state initializers dropped, some are still required as "cleaners" on refresh/startup.
* `syncthing-auto-refresh-mode` for auto-refreshing

progfolio commented:
Good job.
Bear in mind, I have no say over what makes it into MELPA.
I just try to offer advice to keep quality high in the elisp ecosystem.

Some more notes:

"=text=" in some of your docstrings looks like org-mode markup.
See: https://www.gnu.org/software/emacs/manual/html_node/elisp/Documentation-Tips.html

> When a documentation string refers to a Lisp symbol, write it as it would be printed
> (which usually means in lower case), with a grave accent ‘`’ before and apostrophe ‘'’ after it.
> There are two exceptions: write t and nil without surrounding punctuation.

That will allow describe-function to turn the symbols into links.


For your faces, some have good semantic names, e.g. `syncthing-title`.
However, if a user customizes the face `syncthing-green` so that is red, the name no longer makes sense. It looks like this is used to indicate a folder is finished syncing?
How about something like `syncthing-success` or `syncthing-synced`, which can inherit from the standard `success` face? This also raises the question of whether or not 5 colors are necessary to indicate syncing status. 

I also think you can drop all the propertizing functions.
Most of them are only used once and exist to partially apply proprtize with a face.
For exmaple:

```
;;@MAYBE: use less, but more semantically meaningful faces.
;;e.g. syncthing-syncing and syncthing-synced
(defun syncthing--propertize-float (float)
  "Return propertized FLOAT formatted according to `syncthing-float-format'."
  (propertize (format syncthing-format-perc float)
              'face (cond ((< perc 25)      'syncthing-red) 
                          ((>= 25 float 49) 'syncthing-orange) 
                          ((>= 50 float 74) 'syncthing-yellow)
                          ((>= 75 float 99) 'sync-thing-light-green)
                          ((= 100)          'sync-thing-green))))
```


Manually incrementing your client buffer names should not be necessary.
See `generate-new-buffer-name` and/or `generate-new-buffer`.


I have more ideas for how the design could be improved, but I don't want to overwhelm.
Let me know if you're interested.

KeyWeeUsr commented:
> Bear in mind, I have no say over what makes it into MELPA. I just try to offer advice to keep quality high in the elisp ecosystem.

Hehehe, I've noticed, but they were still very nice suggestions and comparing to the initial state the code looks more favorable to me too.

> "=text=" in some of your docstrings looks like org-mode markup.

Yeah, I was trying to avoid that due to that format being just ugly or inconsistent. I much more prefer `<backtick>word<backtick>` (sad GitHub formatting noises) or `=word=` for programming-related text. Sadly, it wasn't picked up by Customize, `describe-*` and friends. Subjective opinion, yet I still think it's a weird choice.

The faces were just me being a bit lazy after the initial restructuring. Now it should look better. The colors are there because if you have large folders (as I do) signification of where exactly it is just by color is quite fast and one (me) can adjust time when syncing between devices which aren't continuously online - aka convenience instead of binary "working/done" state.

I've dropped some propertizing funcs, but left some due to encapsulation. Perhaps I should use `defsubst` for them :thinking: 

> See generate-new-buffer-name and/or generate-new-buffer.

That's quite a nice func! A bit annoying I couldn't wrap it in `*` by default.

> It would be nice if the buffer name were based off of the server's address.

Agreed, I've also added an interactive func that would take such an API base URL in case one wants to open multiple sessions without modifying the default customization value.

> I have more ideas for how the design could be improved, but I don't want to overwhelm.

Yes please, I think it's heading in quite a nice direction. I'll still need to pull the state somewhere else as this elisp-y way of rendering widgets reminds me of React and it's tree rendering, for which the current approach for update/refresh is a bit over the top instead of rather linear. What can one do, another rework, eventually, once I include the rest of the values and writing API, hehehe. But for the initial version of a "I'm too lazy even to open a browser" it works quite nicely. :blush:

riscy commented:
Looks good to merge on this end - thanks both. :)

riscy commented:
One last bit of feedback re: the package metadata is the "Emacs" in "Emacs client for Syncthing" isn't necessary, e.g. this could just read "Client for Syncthing."   In the `;;; Commentary`, a link to the service's page (e.g. syncthing.net?) would be helpful, saving users from having to search and/or disambiguate to understand what this is.
